### PR TITLE
libnetwork: share a single datastore with drivers

### PR DESCRIPTION
**Pull request by corhere**
The bbolt library wants exclusive access to the boltdb file and uses file locking to assure that is the case. The controller and each network driver that needs persistent storage instantiates its own unique datastore instance, backed by the same boltdb file. The boltdb kvstore implementation works around multiple access to the same boltdb file by aggressively closing the boltdb file between each transaction. This is very inefficient. Have the controller pass its datastore instance into the drivers and enable the PersistConnection option to disable closing the boltdb between transactions.

**- What I did**

**- How I did it**

**- How to verify it**
Watch CI and hope for the best

**- Description for the changelog**

**- A picture of a cute animal (not mandatory but encouraged)**

### CODE REVIEW

**Comment by thaJeztah** [comment]
Looks like some failures are consistent between graphdriver and snapshotter runs may be legit; 


```
=== Failed
=== FAIL: daemon TestIpcPrivateVsReadonly (59.98s)
    oci_linux_test.go:111: assertion failed: error is not nil: timeout

=== FAIL: daemon TestSysctlOverride (59.95s)
    oci_linux_test.go:139: assertion failed: error is not nil: timeout

=== FAIL: daemon TestSysctlOverrideHost (60.00s)
    oci_linux_test.go:189: assertion failed: error is not nil: timeout

=== FAIL: daemon TestDefaultResources (59.99s)
    oci_linux_test.go:226: assertion failed: error is not nil: timeout

=== FAIL: daemon TestDaemonReloadNetworkDiagnosticPort (59.97s)
    reload_test.go:369: timeout

=== FAIL: daemon TestGetBlkioWeightDevices (unknown)

=== FAIL: daemon TestGetBlkioThrottleDevices (unknown)

=== FAIL: daemon TestExecResizeTimeout (unknown)
coverage: 13.3% of statements
```

**Comment by thaJeztah** [comment]
No clue why they fail on this PR though 🤔 Let me restart CI

**Comment by corhere** [comment]
Nah, those failures are legit. There's nothing (necessarily) wrong with the code, but the tests need to be changed so as not to instantiate libnetwork controllers with the default datastore path

**Comment by thaJeztah** [comment]
Yeah, I later noticed that one of the tests had a stack related to libnetwork, so perhaps that affects the other tests.

**Comment by thaJeztah** [comment]
Oh! Nice and shiny green now ✅ 🥳 


@akerouanton @robmry ptal 🤗

**Review by robmry — APPROVED** [review]
LGTM

File: libnetwork/datastore/datastore.go [hunk 1]
```diff
@@ -93,6 +92,7 @@ func DefaultScope(dataDir string) ScopeCfg {
 			Config: &store.Config{
 				Bucket:            "libnetwork",
 				ConnectionTimeout: time.Minute,
+				PersistConnection: true,
```
Comment (akerouanton) [inline libnetwork/datastore/datastore.go:95]: Do you plan to open a follow-up PR to remove this option?

File: libnetwork/drivers/bridge/bridge_store.go [hunk 2]
```diff
@@ -25,17 +24,13 @@ const (
 
 func (d *driver) initStore(option map[string]interface{}) error {
 	if data, ok := option[netlabel.LocalKVClient]; ok {
```
Comment (akerouanton) [inline libnetwork/drivers/bridge/bridge_store.go:26]: It'd be cleaner to change drivers' `Register` function to add a dedicated arg for passing the datastore. Is this something you're planning to do in a follow-up?

Comment (corhere) [reply libnetwork/drivers/bridge/bridge_store.go:26]: More type-safe, sure, but that'd just be changing the shape of dependency injection. I don't think it's high priority; I'd like to leave it alone for now and wait for a wider refactoring of the driver interface to make such a change.

File: libnetwork/datastore/datastore.go (hunk 1 above)
Comment (corhere) [reply libnetwork/datastore/datastore.go:95]: Eventually, I guess. Do you want to do it?

**Review by akerouanton — APPROVED** [review]
LGTM

File: libnetwork/datastore/datastore.go (hunk 1 above)
Comment (akerouanton) [reply libnetwork/datastore/datastore.go:95]: - https://github.com/moby/moby/pull/47302

**Pull request merged**